File: DSAlgorithms/patterns/fast_slow_pointers/2_linked_list_cycle_start.py
from  __future__ import print_function

class node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next;
    def print_list(self):
        temp = self
        while temp is not None:
            print(temp.value, end='')
            temp = temp.next
        print()
def find_cycle_start(head):
    cycle_length = 0
    slow, fast = head, head
    while(fast is not None and fast.next is not None):
        fast = fast.next.next
        slow = slow.next
        if slow == fast:
            cycle_length = calc_cycle_length(slow)
            break
    return find_start(head, cycle_length)
def calc_cycle_length(slow):
    current = slow
    cycle_length = 0;
    while True:
        current = current.next
        cycle_length += 1
        if current == slow:
            break
    return cycle_length
def find_start(head, cycle_length):
    pointer_1 = head
    pointer_2 = head
    while cycle_length > 0:
        pointer_2 = pointer_2.next
        cycle_length -= 1
    while pointer_1 != pointer_2:
        pointer_1 = pointer_1.next
        pointer_2 = pointer_2.next
    return pointer_1

def main():
    head = node(1)
    head.next = node(2)
    head.next.next = node(3)
    head.next.next.next = node(4)
    head.next.next.next.next = node(5)
    head.next.next.next.next.next = node(6)

    head.next.next.next.next.next.next = head.next.next
    print("list cycle start: " + str(find_cycle_start(head).value))

    head.next.next.next.next.next.next = head.next.next.next
    print("list cycle start: " + str(find_cycle_start(head).value))
    
    head.next.next.next.next.next.next = head
    print("list cycle start: " + str(find_cycle_start(head).value))

main()

File: DSAlgorithms/patterns/fast_slow_pointers/2_linked_list_cycle_start.cpp
#include <bits/stdc++.h>

class list_node {
    public:
        int value = 0;
        list_node *next;
        list_node(int value)
        {
            this->value = value;
            next = nullptr;
        }
};

class linked_list_cycle_start{
    public:
        static list_node *find_cycle_start(list_node *head)
        {
            int cycle_length = 0;
            list_node *slow = head;
            list_node *fast = head;
            while(fast != nullptr && fast->next != nullptr)
            {
                fast = fast->next->next;
                slow = slow->next;
                if(slow == fast)  
                {
                    cycle_length = calc_cycle_length(slow);
                    break;
                }
            }
            return find_start(head, cycle_length);
        }
    private:
        static int calc_cycle_length(list_node *slow)
        {
            list_node *current = slow;
            int cycle_length = 0;
            do{
                current = current->next;
                cycle_length++;
            }
            while(current != slow);
            return cycle_length;
        }
        static list_node *find_start(list_node *head, int cycle_length)
        {
            list_node *pointer_1 = head;
            list_node *pointer_2 = head;
            while(cycle_length > 0)
            {
                pointer_2 = pointer_2->next;
                cycle_length--;
            }
            while(pointer_1 != pointer_2)
            {
                pointer_1 = pointer_1->next;
                pointer_2 = pointer_2->next;
            }
            return pointer_1;
        }
};

int main(int argc, char *argv[])
{
    list_node *head = new list_node(1);
    head->next = new list_node(2);
    head->next->next = new list_node(3);
    head->next->next->next = new list_node(4);
    head->next->next->next->next = new list_node(5);
    head->next->next->next->next->next = new list_node(6);

    head->next->next->next->next->next->next = head->next->next;
    std::cout<<"list cycle start: "<<linked_list_cycle_start::find_cycle_start(head)->value<<'\n';

    head->next->next->next->next->next->next = head->next->next->next;
    std::cout<<"list cycle start: "<<linked_list_cycle_start::find_cycle_start(head)->value<<'\n';

    head->next->next->next->next->next->next = head;
    std::cout<<"list cycle start: "<<linked_list_cycle_start::find_cycle_start(head)->value<<'\n';
}
